### PR TITLE
add grpcio package

### PR DESCRIPTION
**Pull request by david-gang**
This is a very challenging module with c code and cython. I need it as transitive dependency for google datastore. Let's see if i get it right in the first pass. Feedback is more than welcome

### CODE REVIEW

**Comment by david-gang** [comment]
I need the help of the conda-forge experts :-)

I get the following error in circleci:

Traceback (most recent call last):
  File "/opt/conda/lib/python3.6/site-packages/conda_build/build.py", line 1530, in build_tree
    test(pkg, config=config)
  File "/opt/conda/lib/python3.6/site-packages/conda_build/build.py", line 1336, in test
    create_env(config.test_prefix, specs, config=config)
  File "/opt/conda/lib/python3.6/site-packages/conda_build/build.py", line 675, in create_env
    actions = install_actions(prefix, index, specs)
  File "/opt/conda/lib/python3.6/site-packages/conda/plan.py", line 461, in install_actions
    update_deps, pinned)
  File "/opt/conda/lib/python3.6/site-packages/conda/plan.py", line 634, in get_actions_for_dists
    pkgs = r.install(specs, installed, update_deps=update_deps)
  File "/opt/conda/lib/python3.6/site-packages/conda/resolve.py", line 813, in install
    pkgs = self.solve(specs, returnall=returnall)
  File "/opt/conda/lib/python3.6/site-packages/conda/resolve.py", line 868, in solve
    self.find_conflicts(specs)
  File "/opt/conda/lib/python3.6/site-packages/conda/resolve.py", line 369, in find_conflicts
    raise UnsatisfiableError(bad_deps)
conda.exceptions.UnsatisfiableError: The following specifications were found to be in conflict:
  - grpcio 1.6.3 py36_0
Use "conda info <package>" to see the dependencies for each package.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/conda/bin/conda-build-all", line 6, in <module>
    exit(main())
  File "/opt/conda/lib/python3.6/site-packages/conda_build_all/cli.py", line 93, in main
    b.main()
  File "/opt/conda/lib/python3.6/site-packages/conda_build_all/builder.py", line 277, in main
    built_dist_location = self.build(meta, build_config)
  File "/opt/conda/lib/python3.6/site-packages/conda_build_all/builder.py", line 209, in build
    conda_build.api.build(meta.meta, config=config)
  File "/opt/conda/lib/python3.6/site-packages/conda_build/api.py", line 97, in build
    need_source_download=need_source_download, config=config)
  File "/opt/conda/lib/python3.6/site-packages/conda_build/build.py", line 1549, in build_tree
    .format(recipe) + error_str + "\n" + extra_help)
RuntimeError: Can't build {'undefined_jinja_vars': [], 'config': <conda_build.config.Config object at 0x7f86d58a39b0>, 'meta_path': '/root/conda-recipes/grpcio/meta.yaml', 'path': '/root/conda-recipes/grpcio', 'requirements_path': '/root/conda-recipes/grpcio/requirements.txt', 'meta': {'package': {'name': 'grpcio', 'version': '1.6.3'}, 'source': {'fn': 'grpcio-1.6.3.tar.gz', 'url': 'https://pypi.io/packages/source/g/grpcio/grpcio-1.6.3.tar.gz', 'sha256': 'e9662782c58bc21be26163b78136eaed091dfd45fac699cb711ee9eeeb7e2f9b'}, 'build': {'number': '0', 'script': 'python setup.py install --single-version-externally-managed --record record.txt'}, 'requirements': {'build': ['python', 'setuptools', 'cython'], 'run': ['python', 'six >=1.5.2', 'protobuf >=3.3.0', 'futures >=2.2.0', 'enum34 >=1.0.4']}, 'test': {'imports': ['grpc']}, 'about': {'home': 'http://github.com/simplejson/simplejson', 'license': 'Apache 2.0', 'license_family': 'Apache', 'summary': 'HTTP/2-based RPC framework', 'description': 'simplejson is a simple, fast, complete, correct and extensible\nJSON <http://json.org> encoder and decoder for Python 2.5+ and\nPython 3.3+. It is pure Python code with no dependencies, but includes\nan optional C extension for a serious speed boost.\n', 'doc_url': 'https://grpc.io/docs/quickstart/python.html', 'dev_url': 'https://github.com/grpc/grpc/tree/master/src/python'}, 'extra': {'recipe-maintainers': ['david-gang'], 'final': True}}} due to environment creation error:
The following specifications were found to be in conflict:
  - grpcio 1.6.3 py36_0
Use "conda info <package>" to see the dependencies for each package.

What is weird, is that the module i want to build is grpcio, so why are the specifications in conflict.

File: recipes/grpcio/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+requirements:
+  build:
+    - python
+    - setuptools
+    - cython
+  run:
+    - python
+    - six >=1.5.2
+    - protobuf >=3.3.0
+    - futures >=2.2.0
```
Comment (synapticarbors) [inline recipes/grpcio/meta.yaml]: It seems like `futures` and `enum34` are only required for python2, so you should add the appropriate tags. See example here:

https://conda.io/docs/user-guide/tasks/build-packages/define-metadata.html#preprocessing-selectors

File: recipes/grpcio/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+  imports:
+    - grpc
+
+about:
+  home: https://grpc.io
+  license: Apache 2.0
+  license_family: Apache
+  summary: 'HTTP/2-based RPC framework'
+
+  description: |
```
Comment (synapticarbors) [inline recipes/grpcio/meta.yaml]: Description is leftover from simplejson example.

File: recipes/grpcio/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    - futures >=2.2.0
+    - enum34 >=1.0.4 
+test:
+  imports:
+    - grpc
+
+about:
+  home: https://grpc.io
+  license: Apache 2.0
+  license_family: Apache
```
Comment (synapticarbors) [inline recipes/grpcio/meta.yaml]: Manually package `license_file`:

https://conda-forge.org/docs/meta.html#packaging-the-license-manually

and add issue upstream and copy issue url in the recipe as a comment.

File: recipes/grpcio/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+build:
+  number: 0
+  script: python setup.py install --single-version-externally-managed --record record.txt
+
+requirements:
+  build:
+    - python
+    - setuptools
+    - cython
```
Comment (synapticarbors) [inline recipes/grpcio/meta.yaml]: Add `toolchain` as a `build` requirement.

**Comment by synapticarbors** [comment]
I'm not sure about the errors your seeing. On CircleCI it looks like it's only failing for py36. On Appveyor it's only passing for py35. Maybe others have an idea.

**Comment by synapticarbors** [comment]
You can try asking on gitter:

https://gitter.im/conda-forge/conda-forge.github.io

although typically the staged-recipes PR is also a good place if you can attract responses.

File: recipes/grpcio/meta.yaml (hunk 4 above)
Comment (isuruf) [reply recipes/grpcio/meta.yaml]: @david-gang, please see the comment above

Comment (david-gang) [reply recipes/grpcio/meta.yaml]: @isuruf I fixed the issues and still have errors in circle ci. Thanks for the help

**Comment by msarahan** [comment]
@david-gang that error almost always indicates missing dependencies higher up in the stack.  Conda's error messages definitely need improvement.  It's a known issue.  Conda-build is passing along what it gets from conda.

Newer versions of conda-build (3-series only) try to create the test env prior to actually building anything.  This usually ferrets out missing dependencies sooner, and gives more straightforward output.  If you feel like backporting, this is the code that is doing that: https://github.com/conda/conda-build/blob/master/conda_build/build.py#L969-L995

The best way to sort this out with conda-build 2 is to try to create environments with each dep until you find the one that is missing.  You'll have commands like:

``conda create -n testenv python=3.6 somedep``

I don't generally actually create the env, but it's a good way to test satisfiability of an env.

You could also install conda-build 3 locally, and try to run your builds in a way that imitates the conda-forge builds.

**Comment by jeffkpayne** [comment]
@david-gang BTW, really great to see someone tackling this one!  Been struggling with it off and on for the last couple of days.  Were you able to figure out which dependency is missing by building with `conda-build 3`?  I just cloned your repo, and will be doing that today.  If I find anything I will definitely comment.

**Comment by david-gang** [comment]
Hi @jeffkpayne , This module is really a pain. I tried everything. I succeded compiling it with python3 but not with python2. It dies in the g++ call.

I gave up on it in the meantime and using pip.
Will be happy to hear if you have advance on it.

Thanks

**Pull request closed without merging**